### PR TITLE
2909 create dependency resolver project and publish an artifact

### LINKED ISSUE #2909: Create "dependency resolver project" and publish an artifact
Implement a new resolver project in PK-Sim. It should depend on PKSim.R only.

From there we can retrieve all the dependencies required for OSPSuite-R and publish them as an artifact.

https://github.com/Open-Systems-Pharmacology/PK-Sim/issues/2905

### PR DESCRIPTION
**Pull request by benjaperez1983**
Fixes #2909 

# Description

This PR is for creating a dependency project and publish the artifacts needed to run the last PK-SIM

## Type of change

Please mark relevant options with an `x` in the brackets.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires documentation changes (link at least one [user](https://github.com/Open-Systems-Pharmacology/docs) or [developer](https://github.com/Open-Systems-Pharmacology/developer-docs) documentation issue):
- [ ] Algorithm update - updates algorithm documentation/questions/answers etc.
- [ ] Other (please describe):

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [ ] Integration tests
- [ ] Unit tests
- [x] Manual tests 
- [ ] No tests required

# Reviewer checklist

Mark everything that needs to be checked before merging the PR.

- [ ] Check if the code is well documented
- [ ] Check if the behavior is what is expected
- [ ] Check if the code is well tested
- [ ] Check if the code is readable and well formatted
- [ ] Additional checks (document below if any)
- [ ] Check if documentation update issue(s) are created if the option `This change requires a documentation update` above is selected

# Screenshots (if appropriate):

# Questions (if appropriate):

### CODE REVIEW

File: appveyor.yml [hunk 1]
```diff
@@ -44,10 +44,18 @@ before_build:
   - nuget sources add -name core -source https://ci.appveyor.com/nuget/ospsuite-core
   - nuget restore
 
+
+after_build:
+  - cd src\ConsoleApp\bin\%CONFIGURATION%
+  - 7z a package.zip *
```
Comment (rwmcintosh) [inline appveyor.yml:50]: This should probably be only some of the files. We don't need the PDB files or many of the XML files (we will need OSPSuite.Dimensions.xml and maybe OSPSuite.PKParameters.xml. I don't think we need .config files either if they are there. This line probably zips up the .exe file too which we don't need.

Comment (rwmcintosh) [reply appveyor.yml:50]: Let's name the package something like `pk-sim-r-dependencies.zip`

File: src/ConsoleApp/Properties/AssemblyInfo.cs
Comment (rwmcintosh) [inline src/ConsoleApp/Properties/AssemblyInfo.cs:1]: Since you created are using SolutionInfo.cs can this be deleted?

File: src/ConsoleApp/PKDependencyApp.csproj
Comment (rwmcintosh) [inline src/ConsoleApp/PKDependencyApp.csproj:1]: Similar to the package name let's rename this to say what it is. `PKSimRDependencyResolution` or something.

File: src/ConsoleApp/Program.cs [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+using System.Text;
+using System.Threading.Tasks;
+
+namespace ConsoleApp
+{
+   internal class Program
+   {
+      static void Main(string[] args)
+      {
+         Console.WriteLine("Dummy Message");
```
Comment (rwmcintosh) [inline src/ConsoleApp/Program.cs:13]: We might as well make the text say what the purpose of the project is. Describe that there's no functionality, the project is used to resolve dependencies for R or something.

**Review by rwmcintosh — CHANGES_REQUESTED** [review]
Just a few comments to polish it up a bit. Looks simple and we already know it works great.

**Comment by rwmcintosh** [comment]
Fill in the #2909 (without space) there so that when the PR is merged, the issue is closed. Also all linking between PR and issue is made using that #

image

**Comment by benjaperez1983** [comment]
@PavelBal  @Yuri05 @msevestre

**Comment by rwmcintosh** [comment]
> Shall we run the build and publish the artifact with each merge? or whall we do it on the nightly build?

We would only see new Core updates propagated to OSPSuite-R after a nightly, rather than on-demand. I think it's not bad if Core updates in OSPSuite-R are nightly.

File: appveyor.yml [hunk 3]
```diff
@@ -3,6 +3,9 @@ image: Visual Studio 2022
 
 version: '{build}'
 
+environment:
+  TARGET_FRAMEWORK: net472
```
Comment (rwmcintosh) [inline appveyor.yml:7]: Why is this needed?

**Review by rwmcintosh — APPROVED** [review]
Just a comment on the environment variable being set in AppVeyor. I'm surprised by  that. Can you comment?

File: appveyor.yml (hunk 3 above)
Comment (msevestre) [reply appveyor.yml:7]: same question

File: src/PKSimRDependencyResolution/PKSimRDependencyResolution.csproj [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+    <PackageReference Include="OSPSuite.FuncParser" Version="4.0.0.54" />
+    <PackageReference Include="OSPSuite.Infrastructure" Version="12.0.296" />
+    <PackageReference Include="OSPSuite.Infrastructure.Autofac" Version="12.0.296" />
+    <PackageReference Include="OSPSuite.Infrastructure.Import" Version="12.0.296" />
+    <PackageReference Include="OSPSuite.R" Version="12.0.296" />
+    <PackageReference Include="OSPSuite.Serializer" Version="3.0.0.1" />
+    <PackageReference Include="OSPSuite.SimModel" Version="4.0.0.56" />
+    <PackageReference Include="OSPSuite.Utility" Version="4.1.0.5" />
+    <PackageReference Include="Portable.BouncyCastle" Version="1.8.9" />
+    <PackageReference Include="Serilog" Version="2.10.0" />
```
Comment (msevestre) [inline src/PKSimRDependencyResolution/PKSimRDependencyResolution.csproj:54]: Are all of these required or can we somehow reference InfrasturacutreImport that depends on these already? Why is this package depending on so many DIRECTLY?

File: src/PKSimRDependencyResolution/Program.cs [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+using System.Text;
+using System.Threading.Tasks;
+
+namespace ConsoleApp
+{
+   internal class Program
+   {
+      static void Main(string[] args)
+      {
+         Console.WriteLine("This console application includes all the dependencies needed for the PK-SIM project");
```
Comment (msevestre) [inline src/PKSimRDependencyResolution/Program.cs:13]: `PK-Sim` !

File: src/PKSimRDependencyResolution/filelist.txt [hunk 6]
```diff
@@ -0,0 +1,62 @@
+Autofac.pdb
```
Comment (msevestre) [inline src/PKSimRDependencyResolution/filelist.txt:1]: I don't get waht this file is? Why do we need this?
And can we generate it automatically instead of statically?

Comment (msevestre) [reply src/PKSimRDependencyResolution/filelist.txt:1]: why are we copying all xmls?

Comment (msevestre) [reply src/PKSimRDependencyResolution/filelist.txt:1]: why are we copying all pdbs?

**Review by msevestre — COMMENTED** [review]
some questions remain

File: appveyor.yml (hunk 3 above)
Comment (benjaperez1983) [reply appveyor.yml:7]: Sure, since yesterday, we updated the PKSimRDependencyResolution.cproj file, remember?
Before this update, the folder containing the files was "src\PKSimRDependencyResolution\bin\debug\"
But now, I can see (both the pipeline and local, of course) the folder has changed to "src\PKSimRDependencyResolution\bin\debug\net472\", so this is the folder to ZIP and distribute.
as net472 is the target framework, it did nos seem like a good idea to hardcode that value, but better set it on a variable.

File: src/PKSimRDependencyResolution/filelist.txt (hunk 6 above)
Comment (benjaperez1983) [reply src/PKSimRDependencyResolution/filelist.txt:1]: It is the opposite in fact, this file contains all the files that we want to EXCLUDE, this is the way we achieve that:
- 7z a pk-sim-r-dependencies.zip -x@filelist.txt * --> "-x" does the trick.
The thing is, we need to exclude certain files, but we do not know in advance which ones, given that we need some .xms for example.
So the approach here is that if we were to include some dependencies in the future, and forgot to update this file, the worst case scenario will be having a zip with more files than we need, and not otherwise, which could cause not to have the files needed for it to run properly

File: src/PKSimRDependencyResolution/filelist.txt
Comment (Yuri05) [inline src/PKSimRDependencyResolution/filelist.txt:1]: Would rename the file to e.g. _**exclude**filelist.txt_ for clarity.

File: appveyor.yml (hunk 3 above)
Comment (rwmcintosh) [reply appveyor.yml:7]: ah yes, I didn't see how it was being used below. It's fine for me, and I would also have accepted it as hard-coded in the path.

In other words, I don't have a strong preference.

**Comment by rwmcintosh** [comment]
@Yuri05 @msevestre @benjaperez1983 @PavelBal still a question whether we should trigger this artifact nightly, or on each PR.

I think it belongs nightly...

**Comment by Yuri05** [comment]
>I think it belongs nightly...

Agree.

File: src/PKSimRDependencyResolution/Program.cs [hunk 7]
```diff
@@ -0,0 +1,16 @@
...
+using System.Text;
+using System.Threading.Tasks;
+
+namespace ConsoleApp
+{
+   internal class Program
+   {
+      static void Main(string[] args)
+      {
+         Console.WriteLine("This console application includes all the dependencies needed for PK-SIM");
```
Comment (Yuri05) [inline src/PKSimRDependencyResolution/Program.cs:13]: **PK-Sim** instead of **PK-SIM** 😉

**Comment by Yuri05** [comment]
What just comes into my mind: do we need separate packages for Windows and Linux?
At the moment the package contains Windows versions of all C++ libs.
Is probably ok for now, but latest when we have migrated PK-Sim Core/Infrastructure to .net8 and going to use the package under Linux as well, we would need 2 separate package versions.

**Comment by rwmcintosh** [comment]
> What just comes into my mind: do we need separate packages for Windows and Linux?

I think so, but maybe for this iteration we can consume the package as built by AppVeyor and replace the native dll as we are now.

File: src/PKSimRDependencyResolution/excludedFiles.txt [hunk 8]
```diff
@@ -0,0 +1,62 @@
+Autofac.pdb
+Autofac.xml
+BouncyCastle.Crypto.xml
+ConsoleApp.exe
+ConsoleApp.exe.config
+ConsoleApp.pdb
+DevExpress.Data.v21.2.xml
+filelist.txt
```
Comment (Yuri05) [inline src/PKSimRDependencyResolution/excludedFiles.txt:8]: This should become **excludedFiles.txt** - right? :)

**Comment by Yuri05** [comment]
Apart from the missing Linux build and https://github.com/Open-Systems-Pharmacology/PK-Sim/pull/2915/files?w=1#r1672413911 - looks good for me :)

**Comment by Yuri05** [comment]
and yes, as mentioned above, it should be actually part of the **nigthly** build

**Comment by rwmcintosh** [comment]
> Apart from the missing Linux build and https://github.com/Open-Systems-Pharmacology/PK-Sim/pull/2915/files?w=1#r1672413911 - looks good for me :)

The Linux build piece will be part of the OSPSuite-R PR, though, right, as it is now?

**Comment by Yuri05** [comment]
> > Apart from the missing Linux build and https://github.com/Open-Systems-Pharmacology/PK-Sim/pull/2915/files?w=1#r1672413911 - looks good for me :)
> 
> The Linux build piece will be part of the OSPSuite-R PR, though, right, as it is now?

hmm, maybe

**Review by Yuri05 — APPROVED** [review]
